src/gtk/overlay.cpp

```diff
@@ -0,0 +1,222 @@
+// wxOverlay
+// ----------------------------------------------------------------------------
+
+wxOverlayImpl::wxOverlayImpl()
+{
+    m_surface = NULL;
+    m_window = NULL;
+    m_cairoContext = NULL;
+    m_cairoSurface = NULL;
+}
+
+wxOverlayImpl::~wxOverlayImpl()
+{
+    if ( m_window )
+        m_window->Unbind(wxEVT_SIZE, &wxOverlayImpl::OnSize, this);
+
+    if ( m_cairoSurface )
+        cairo_surface_destroy(m_cairoSurface);
+
+    if ( m_surface )
+    {
+        gtk_widget_destroy(m_surface);
+        g_object_unref(m_surface);
+    }
+}
+
+bool wxOverlayImpl::IsOk() const
+{
+    return m_window != NULL;
+}
+
+void wxOverlayImpl::Init(wxDC *dc, int x, int y, int width, int height)
+{
+    wxCHECK_RET( dc, "NULL dc pointer" );
+    wxASSERT_MSG( !IsOk() , "You cannot Init an overlay twice" );
+
+    wxLogDebug("wxOverlayImpl::Init(%d, %d, %d, %d)", x, y, width, height);
+
+    m_window = dc->GetWindow();
+    m_window->Bind(wxEVT_SIZE, &wxOverlayImpl::OnSize, this);
+
+    m_rect = wxRect(x, y, width, height);
+
+    int clientWidth, clientHeight;
+    m_window->GetClientSize(&clientWidth, &clientHeight);
+
+    if ( m_surface )
+    {
+        gtk_widget_set_size_request(m_surface, clientWidth, clientHeight);
+
+        return;
+    }
+
+    // Create the surface:
+
+    GtkWidget* const overlay = GetOverlayWidget(m_window);
+    wxASSERT_MSG( overlay, "Overlay is not enabled for m_window" );
+
+    m_surface = gtk_drawing_area_new();
+    gtk_widget_set_has_window(m_surface, false);
+    gtk_widget_set_size_request(m_surface, clientWidth, clientHeight);
+    gtk_widget_set_halign (m_surface, GTK_ALIGN_START);
+    gtk_widget_set_valign (m_surface, GTK_ALIGN_START);
+    gtk_widget_set_vexpand (m_surface, TRUE);
+    gtk_widget_set_hexpand (m_surface, TRUE);
+    gtk_widget_set_can_focus(m_surface, false);
+
+    gtk_overlay_add_overlay( GTK_OVERLAY(overlay), m_surface );
+    gtk_overlay_set_overlay_pass_through( GTK_OVERLAY(overlay), m_surface, true );
+
+    g_signal_connect(m_surface, "draw", G_CALLBACK(wxgtk_draw_overlay), this);
+
+    g_object_ref(m_surface);
+    gtk_widget_show(m_surface);
+}
+
+void wxOverlayImpl::BeginDrawing(wxDC *dc)
+{
+    wxCHECK_RET( dc, "NULL dc pointer" );
+
+    wxClientDCImpl* const
+        dcimpl = static_cast<wxClientDCImpl*>(dc->GetImpl());
+
+    wxCHECK_RET( dcimpl, wxS("DC is not a wxClientDC!") );
+
+    if ( m_cairoSurface )
+        cairo_surface_destroy(m_cairoSurface);
+
+    m_cairoContext = static_cast<cairo_t*>(dcimpl->GetCairoContext());
+    cairo_push_group(m_cairoContext);
+
+    m_cairoSurface = cairo_get_group_target(m_cairoContext);
+    cairo_surface_reference(m_cairoSurface);
+}
+
+void wxOverlayImpl::EndDrawing(wxDC *WXUNUSED(dc))
+{
+    if ( m_cairoContext )
+    {
+        cairo_pop_group_to_source(m_cairoContext);
+        cairo_surface_set_device_offset(m_cairoSurface, 0, 0);
+
+        m_window->RefreshRect(m_rect);
```

Comment:
If the window is going to be re-painted, I don't understand the point of using `GtkOverlay`. Why not just draw an overlay surface at the end of the paint?